Fix infinite loop when `posix_spawn` fails (re: f762a81a)

JohnoKing opened the pull request:
This commit fixes an infinite loop introduced in commit f762a81a that caused ksh to enter an infinite loop if `posix_spawn` failed
to start a new process after setting the terminal process group. Reproducer (warning: it will cause ksh to crash Wayland sessions and drives up CPU usage by a ton):
```sh
$ /tmp/this/file/does/not/exist
/usr/bin/ksh: /tmp/this/file/does/not/exist: not found
$ # Press enter
$ 
# ksh now prints $PS1 in a loop until killed with SIGKILL
```

src/cmd/ksh93/sh/xec.c: `sh_ntfork()`:
- Restore the terminal process group if `posix_spawn` failed to launch a new process. This is necessary because `posix_spawn` will set the terminal process group before it attempts to run a command and doesn't restore it on failure.

JohnoKing committed:
Fix infinite loop when posix_spawn fails (re: f762a81a)

This commit fixes an infinite loop introduced in commit f762a81a
that caused ksh to enter an infinite loop if posix_spawn failed
to start a new process after setting the terminal process group.
Reproducer (warning: it will cause ksh to crash Wayland sessions
and drives up CPU usage by a ton):
   $ /tmp/this/file/does/not/exist
   /usr/bin/ksh: /tmp/this/file/does/not/exist: not found
   $ <Press enter>
   (ksh now prints $PS1 in a loop until killed with SIGKILL)

src/cmd/ksh93/sh/xec.c: sh_ntfork():
- Restore the terminal process group if posix_spawn failed to launch
  a new process. This is necessary because posix_spawn will set the
  terminal process group before it attempts to run a command and
  doesn't restore it on failure.

JohnoKing commented:
I've found that there is still a problem. For some reason the bugfix is ineffective for commands starting with a slash. Reproducer:
```sh
$ exec arch/*/bin/ksh
$ /
/usr/bin/ksh: /: cannot execute [Is a directory]
$ # Press enter
```
~At this point I'm not sure if this is a bug in ksh or glibc's implementation of `posix_spawn_file_actions_addtcsetpgrp_np`.~

**Edit:** I've pushed an updated version of the infinite loop fix. The new version should fix the remaining problem, though I'd prefer to test it for a while before it's merged upstream.

JohnoKing committed:
Fix SIGTTOU freeze when run with `exec strace`

JohnoKing committed:
Merge remote-tracking branch 'upstream/dev' into fix-tcsetpgrp-loop

JohnoKing committed:
Experimental fix: Set the terminal process group to sh.pid

McDutchie commented:
OK, holding off on merging this on your request. Please let me know when you're ready.

JohnoKing committed:
Merge remote-tracking branch 'upstream/dev' into fix-tcsetpgrp-loop

JohnoKing committed:
Merge remote-tracking branch 'upstream/dev' into fix-tcsetpgrp-loop

JohnoKing committed:
Add NEWS entries for fixed bugs

This commit message will provide explanations for how each bugfix
applied works.

The first bug fixed is the infinite loop that occurs when posix_spawn
fails to execute a command. This was fixed by setting the terminal
process group to the main interactive shell.

The second bug fixed is related to the signal handling of the SIGTTIN,
SIGTTOU and SIGTSTP signals. In sh_ntfork() these signals are set to
their default signal handlers (SIG_DFL) before running a command. The
signal handlers were only restored to SIG_IGN (ignore signal) when
sh_ntfork() successfully ran a command. This could cause a SIGTTOU
lockup under strace when a command failed to execute in an interactive
shell, while also being one cause of the infinite loop.

JohnoKing commented:
I have been using the current patch on my system for a while now (with glibc 2.35) and haven't encountered any additional problems. I've pushed a new commit that adds two entries to the NEWS file and (in the commit message) details how each bugfix in `sh_ntfork()` works. The pull request should be ready to merge now.
